selfhost/typechecker.jakt

```diff
@@ -1460,6 +1548,334 @@ struct Typechecker {
+
+        loop {
+            let fixed_point = .substitute_typevars_in_type_helper(type_id, generic_inferences)
+
+            if fixed_point.equals(result) {
+                break
+            } else {
+                result = fixed_point
+            }
+        }
+        return result
+    }
+
+    // FIXME: Use [TypeId: TypeID] without TypeId.to_string()/TypeId::from_string() workaround
+    function substitute_typevars_in_type_helper(mut this, type_id: TypeId , generic_inferences: [String: String]) throws -> TypeId {
+        let type_ = .get_type(type_id)
+
+        match type_ {
+            TypeVariable() => {
+                let replacment_type_id_string = generic_inferences.get(type_id.to_string())
+                if replacment_type_id_string.has_value() {
+                    return TypeId::from_string(replacment_type_id_string.value())
+                }
+            }
+            GenericInstance(id, args) => {
+                mut new_args = args
+                for arg in new_args.iterator() {
+                    arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
+                }
+
+                return .find_or_add_type_id(Type::GenericInstance(id, args: new_args))
+            }
+            GenericEnumInstance(id, args) => {
+                mut new_args = args
+                for arg in new_args.iterator() {
+                    arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
+                }
+                return .find_or_add_type_id(Type::GenericEnumInstance(id, args: new_args))
+            }
+            Struct(struct_id) => {
+                let struct_ = .get_struct(struct_id)
+                if not struct_.generic_parameters.is_empty() {
+                    mut new_args = struct_.generic_parameters
+                    for arg in new_args.iterator() {
+                        arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
+                    }
+                    return .find_or_add_type_id(Type::GenericInstance(id: struct_id, args: new_args))
+                }
+            }
+            Enum(enum_id) => {
+                let enum_ = .get_enum(enum_id)
+                if not enum_.generic_parameters.is_empty() {
+                    mut new_args = enum_.generic_parameters
+                    for arg in new_args.iterator() {
+                        arg = .substitute_typevars_in_type(type_id: arg, generic_inferences)
+                    }
+                    return .find_or_add_type_id(Type::GenericEnumInstance(id: enum_id, args: new_args))
+                }
+            }
+            RawPtr(rawptr_type_id) => {
+                let rawptr_type = Type::RawPtr(
+                    .substitute_typevars_in_type(type_id: rawptr_type_id, generic_inferences)
+                )
+                return .find_or_add_type_id(rawptr_type)
+            }
+            else => {
+                return type_id
```

Comment:
You could just leave that blank (`else => {}`) as you'd return `type_id` immediately after anyway